languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -163,6 +162,113 @@ def user_in_role_query(
     return query
 
 
+def list_filter_query(kind, resource, relationships, id_field):
+    tablename = resource.python_class.__tablename__
+
+    # Get relationships sorted out.
+    rels = []
+    resource_type = resource.type
+    found_parent = True
+    while found_parent:
+        found_parent = False
+        for rel in relationships:
+            if rel.child_type == resource_type:
+                rels.append(rel)
+                resource_type = rel.parent_type
+                found_parent = True
+                break
+
+    sql = ""
+    assert kind in ["role_allow", "user_in_role"]
+    # Get the roles to start from.
+    if kind == "role_allow":
+        # Any role with the permission.
+        sql += """
+        with allow_permission as (
+            select
+                p.id
+            from permissions p
+            where p.resource_type = :resource_type and p.name = :action
+        ), starting_roles as (
+            select
+                rp.role as role
+            from role_permissions rp
+            join allow_permission ap
+            on rp.permission_id = ap.id
+        ),
+        """
+    elif kind == "user_in_role":
+        # The passed in role.
+        sql += """
+        with starting_roles as (
+            select
+                r.name as role
+            from roles r
+            where r.name = :role
+        ),
+        """
+    # Get to any of the roles the user could be assigned to.
+    sql += """
+    relevant_roles as (
+        with recursive relevant_roles (role) as (
+            select
+                    role
+            from starting_roles
+            union
+            select
+                    ri.from_role
+            from role_implications ri
+            join relevant_roles rr
+            on ri.to_role = rr.role
+        ) select * from relevant_roles
+    ), user_relevant_roles as (
+        select
+            resource_type, resource_id
+        from user_roles ur
+        join relevant_roles rr
+        on rr.role = ur.role
+        where ur.user_id = :user_id
+    )
+    """
+
+    # Select data
+    sql += f"""
+    select
+      {tablename}.{id_field}
+    from {tablename}
+    join user_relevant_roles urr
+    on urr.resource_type = '{resource.type}' and urr.resource_id = {tablename}.{id_field}
+    """
+
+    prev_joins = []
+
+    for i, rel in enumerate(rels):
+        parent_pk, _ = get_pk(rel.parent_python_class)
+
+        join = f"join {rel.parent_table} "
+        join += f"on {rel.child_table}.{rel.child_join_column} "
+        join += f"= {rel.parent_table}.{rel.parent_join_column}"
+        prev_joins.append(join)
+
+        sql += f"""
+        union
+        select
+        {tablename}.{id_field}
+        from {tablename}
+        """
+
+        for join in prev_joins:
```

Comment:
Are there tests that go across > 2 levels of relationships and use data filtering?

Reply:
There's Issue list filtering tests that use 2 relationships